### ISSUE TITLE
performance issues while fetching vectors by ids

### ISSUE THREAD

**Issue by suvashishtha** [milvus-io/milvus#3243]
I am experiencing performance issues while fetching vectors by Id. I am noticing it to be very slow. 

**Steps/Code to reproduce behavior**
- insert 500K vectors
- create index (i tried annoy & ivfsq8)
- try client.getEntitiesById() for 1000 ids

**Expected behavior**
Should respond in some milliseconds

**Environment details**
- Hardware/Softward conditions: MacOS 16G RAM
- Method of installation - docker (6G allocated to docker)
- Milvus version - v0.10.1

Is it a normal behaviour?

**Comment by sgaseretto** [comment]
How much time took milvus to fetch the results?

**Comment by suvashishtha** [comment]
took around 23 seconds with java client. performance was slightly better with rest api (200 millis for 150 records) but couldn't fetch more than 150 for some reason.
Exception i got was

```
I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://localhost:19121: The target server failed to respond
```

**Comment by yhmo** [comment]
Use python sdk to test:
1. insert 500k 128 dimensional vectors
2. create index IVFSQ8
3. invoke get_entity_by_id to input 1000 ids in one time, time cost = 1.1 seconds
4. invoke get_entity_by_id 1000 times, each time fetch one id, time cost = 3.1 seconds

Will do more experiment for rest api

**Comment by yhmo** [comment]
Use curl to send http request to get vectors by id:

`for ((a=1;a<=1000;a+=1)); do curl -X GET "http://192.168.1.20:19121/collections/TEST/vectors?ids=1597284173570220000"; done`

Time cost: 20 ~ 25 seconds
The bottle neck is not in Milvus server side, most of time waste in data serialized/deserialized/printing.

**Comment by yhmo** [comment]
Use curl to send http request to get vectors by id batch(each batch has 50 ids), totally get 1000 ids:

`for ((a=1;a<=20;a+=1)); do curl -X GET "http://192.168.1.20:19121/collections/TEST/vectors?ids=1597284173570220000,1597284173570220001,1597284173570220002,1597284173570220003,1597284173570220004,1597284173570220005,1597284173570220006,1597284173570220007,1597284173570220008,1597284173570220009,1597284173570220010,1597284173570220011,1597284173570220012,1597284173570220013,1597284173570220014,1597284173570220015,1597284173570220016,1597284173570220017,1597284173570220018,1597284173570220019,1597284173570220020,1597284173570220021,1597284173570220022,1597284173570220023,1597284173570220024,1597284173570220025,1597284173570220026,1597284173570220027,1597284173570220028,1597284173570220029,1597284173570220030,1597284173570220031,1597284173570220032,1597284173570220033,1597284173570220034,1597284173570220035,1597284173570220036,1597284173570220037,1597284173570220038,1597284173570220039,1597284173570220040,1597284173570220041,1597284173570220042,1597284173570220043,1597284173570220044,1597284173570220045,1597284173570220046,1597284173570220047,1597284173570220048,1597284173570220049"; done`

Time cost: about 2 seconds

**Comment by suvashishtha** [comment]
Thank you for your response @yhmo . Even with python sdk or curl, fetching 1000 vectors by id taking 1 - 2 seconds still looks to be on slower side, no? Any reason for this slowness?

**Comment by yhmo** [comment]
To fetch a vector by its id: 
Firstly Milvus use bloom filter to quickly find out which segment the vector is possible belong. The bloom filter file in under segment folder, its size is about 2.8MB
Secondly, Milvus load segment's uid file, the id file store all vectors id of this segment and there offset in raw file, under the segment folder, its size is 12*vector_count bytes. by this uid file, milvus know the exactly position of the vector.
Finally, Milvus read the vector from the raw file and return to client.

Milvus doesn't cache the bloom filter file and uid file in memory, so each time you invoke get_entity_by_id, it always read them from disk, this is why the performance is bad for fetching vectors by id.

**Comment by yhmo** [comment]
Seems curl has size limitation for GET, you can search "curl get size limit" by Google for more information.

**Comment by suvashishtha** [comment]
Thanks for the explanation @yhmo. Looks good from my side.

**Comment by suvashishtha** [comment]
I guess fetch vectors by is only a get method, right? Is there a post method also present? If not, will it be ok to add one?

**Comment by yhmo** [comment]
The POST method has been added for insert. The rest api README:
https://github.com/milvus-io/milvus/blob/master/core/src/server/web_impl/README.md

**Comment by suvashishtha** [comment]
Sorry, what i meant is there a post method to fetch vectors by id? The one in readme file is used to insert vectors i guess, right?

**Comment by yhmo** [comment]
Currently we have no plan to add a POST method to fetch vectors by id. Personally I think this is a good point to improve the usability of rest api. But this need to be approved by milvus technical committee, they will collection important issues to discuss before each iteration.

**Comment by suvashishtha** [comment]
All good @yhmo . Can this be added as a low priority enhancement/feature then which is to be considered?

**Comment by yhmo** [comment]
I have created issue #3249 for new rest POST api to fetch vectors by id.

**Comment by suvashishtha** [comment]
Thank you @yhmo. can close this one then

**Comment by suvashishtha** [comment]
one more question please,
when milvus reads raw file to get the vector, does it keep the segment of the file/some ids in cache forever?

**Comment by yhmo** [comment]
No, it doesn't keep ids in cache, neither the vector data read from raw file. 
Milvus only cache data required by search action, raw file or index file is cached in memory entirely. The get_entity_by_id only read some part of the raw file.
